File: src/leetcode/Javascript/217.Contains-Duplicate.js
/*
Given an integer array nums,return true if any value appears at least twice in the array, 
and return false if every element is distinct.
*/

//Brute force solution Time Complexity: O(n^2), Space Complexity: O(1)
const containsDuplicate = (nums) => {
  if (nums.length <= 0) return false;
  for (let i = 0; i < nums.length; i++) {
    for (let j = i + 1; j < nums.length; j++) {
      if (nums[i] === nums[j]) {
        return true;
      }
    }
  }
  return false;
};

// Optimal solution 1, using sorting Time Complexity: O(nlogn), Space Complexity: O(1)
const containDuplicate = (nums) => {
  if (nums.length <= 1) return false;
  nums.sort((a, b) => a - b);
  for (let i = 0; i < nums.length; i++) {
    if (nums[i] === nums[i + 1]) {
      return true;
    }
  }
  return false;
};

// Optimal solution 2, using hashmaps Time Complexity: O(n), Space Complexity: O(n)
const containDuplicate1 = (nums) => {
  let map = new Map();
  let num;
  for (num of nums) {
    if (map.has(num)) return true;
    map.set(num);
  }
  return false;
};

//Optimal solution 3, using sets Time Complexity: O(n), Space Complexity: O(n)
//A set forms a new data type with no duplicates and is unordered
const containDuplicate2 = (nums) => {
  //returns true if the length of the array is not equal to the size of the set created
  return nums.length !== new Set(nums).size

}
//tests
console.log(containsDuplicate([1, 2, 3, 1])); //Output:  true
console.log(containDuplicate([1, 2, 3, 4])); //Output:   false
console.log(containsDuplicate1([1, 1, 1, 3, 3, 4, 3, 2, 4, 2])); //Output:   true
console.log(containsDuplicate1([])); //Output:  false

File: src/leetcode/python/217.Contains-Duplicate.py
'''Given an integer array nums, return true if any value appears at least twice in the array, 
    and return false if every element is distinct.
 '''

class Solution:
    #Brute Force/Naive Approach--> Time complexity O(n^2) Space Complexity O(1)
    def containsDuplicate(self, nums: List[int]) -> bool:
        #Loops from the first element of the array
        for i in range(len(nums)):
            #loops from the second element of the array
            for j in range(i+1, len(nums)):
                if nums[i] == nums[j]:
                    return True 
        return False


    #Optimal sln1-Sorting function: Time Complexity O(n log n) Space complexity O(1)
    def containsDuplicate1(self, nums: List[int]) -> bool:
        if len(nums) < 2: return False
        #sort the array
        nums.sort()
        #The loop runs till the second last element for comparison to occur
        for i in range(len(nums)-1):
            #compare the current element with the next
            if nums[i] == nums[i+1]:
                return True
        return False


    #Optimal sln2 -Use sets : Time Complexity -> O(n) Space Complexity O(n)
    #A set is a data type in python that is unordered, unchangeable*, unindexed and does not contain duplicates
    def containsDuplicate2(self, nums: List[int]) -> bool:
        # this line returns true if the length of the original array 
        #is not equal to the newly formed set because a set has no duplicates
        return len(nums) != len(set(nums))


    #Optimal sln3- Use HashMap: Time Complexity O(n) Space Complexity O(n)
    def containsDuplicate3(self, nums: List[int]) -> bool:
        #hashMap to store the elements traversed
        hashMap = {}        
        for num in  nums:
            #if the element is in the hashMap return true else insert the element in the hashMap
            if num in hashMap:
                return True
            else:
                hashMap[num] = num
        return False


#Test CASES
print(Solution().containsDuplicate3([1,2,3,1])) #True
print(Solution().containsDuplicate3([1,2,3,4])) #False
print(Solution().containsDuplicate3([1,1,1,3,3,4,3,2,4,2])) #True